File: android/app/src/main/kotlin/com/timecoin/wallet/network/ConfigManager.kt
package com.timecoin.wallet.network

import android.util.Log
import java.io.File

/**
 * Manages the `time.conf` configuration file — Bitcoin-style key=value format
 * matching the desktop wallet's config_new.rs.
 *
 * Testnet config lives in `testnet/time.conf` subdirectory;
 * mainnet config lives in `time.conf` at the root. Directory location
 * determines the network — no `testnet=0|1` flag needed.
 *
 * Supported keys:
 *   addnode=IP:PORT   (repeatable — manual masternode peers)
 *   rpcuser=...       (RPC username)
 *   rpcpassword=...   (RPC password)
 */
object ConfigManager {
    private const val TAG = "ConfigManager"
    private const val CONFIG_FILENAME = "time.conf"
    private const val TESTNET_SUBDIR = "testnet"
    private const val MAINNET_PORT = 24001
    private const val TESTNET_PORT = 24101

    data class Config(
        val peers: List<String> = emptyList(),
        val rpcUser: String? = null,
        val rpcPassword: String? = null,
        val testnet: Boolean = false,
    )

    /** Resolve the config directory: testnet uses `testnet/` subdirectory. */
    private fun configDir(baseDir: File, isTestnet: Boolean): File {
        return if (isTestnet) File(baseDir, TESTNET_SUBDIR) else baseDir
    }

    /** Load configuration from time.conf. Testnet reads from testnet/ subdirectory. */
    fun load(dir: File, isTestnet: Boolean = false): Config {
        val file = File(configDir(dir, isTestnet), CONFIG_FILENAME)
        if (!file.exists()) return Config(testnet = isTestnet)
        return try {
            parse(file.readText()).copy(testnet = isTestnet)
        } catch (e: Exception) {
            Log.w(TAG, "Failed to parse $CONFIG_FILENAME: ${e.message}")
            Config(testnet = isTestnet)
        }
    }

    /** Save configuration to time.conf. Testnet saves to testnet/ subdirectory. */
    fun save(dir: File, config: Config) {
        val cfgDir = configDir(dir, config.testnet)
        cfgDir.mkdirs()
        val file = File(cfgDir, CONFIG_FILENAME)
        file.writeText(serialize(config))
        Log.d(TAG, "Saved $CONFIG_FILENAME with ${config.peers.size} peers (testnet=${config.testnet})")
    }

    /** Get the raw text content of time.conf (for editing in the UI). */
    fun readRaw(dir: File, isTestnet: Boolean = false): String {
        val file = File(configDir(dir, isTestnet), CONFIG_FILENAME)
        return if (file.exists()) file.readText() else defaultConfig(isTestnet)
    }

    /** Write raw text content to time.conf. */
    fun writeRaw(dir: File, content: String, isTestnet: Boolean = false) {
        val cfgDir = configDir(dir, isTestnet)
        cfgDir.mkdirs()
        File(cfgDir, CONFIG_FILENAME).writeText(content)
    }

    /** Build endpoint URLs from the addnode entries, adding scheme and default port. */
    fun manualEndpoints(config: Config): List<String> {
        val port = if (config.testnet) TESTNET_PORT else MAINNET_PORT
        return config.peers.map { peer ->
            when {
                peer.startsWith("http://") || peer.startsWith("https://") -> peer
                peer.contains(':') -> "http://$peer"
                else -> "http://$peer:$port"
            }
        }
    }

    /** RPC credentials from config, or null if not set. */
    fun rpcCredentials(config: Config): Pair<String, String>? {
        val user = config.rpcUser ?: return null
        val pass = config.rpcPassword ?: return null
        return user to pass
    }

    // ── Parsing ──

    internal fun parse(contents: String): Config {
        val peers = mutableListOf<String>()
        var rpcUser: String? = null
        var rpcPassword: String? = null

        for (raw in contents.lines()) {
            // Strip comments
            val line = raw.indexOf('#').let { if (it >= 0) raw.substring(0, it) else raw }.trim()
            if (line.isEmpty()) continue
            val eq = line.indexOf('=')
            if (eq < 0) continue
            val key = line.substring(0, eq).trim()
            val value = line.substring(eq + 1).trim()

            when (key) {
                "addnode" -> if (value.isNotEmpty()) peers.add(value)
                "rpcuser" -> if (value.isNotEmpty()) rpcUser = value
                "rpcpassword" -> if (value.isNotEmpty()) rpcPassword = value
                // testnet key is ignored — directory determines network
            }
        }
        return Config(peers, rpcUser, rpcPassword)
    }

    internal fun serialize(config: Config): String = buildString {
        appendLine("# TIME Coin Wallet Configuration")
        appendLine("# Edit this file to add masternode peers manually.")
        appendLine()
        appendLine("# Masternode peers (one per line)")
        if (config.peers.isEmpty()) {
            val examplePort = if (config.testnet) TESTNET_PORT else MAINNET_PORT
            appendLine("# addnode=64.91.241.10:$examplePort")
        } else {
            for (peer in config.peers) {
                appendLine("addnode=$peer")
            }
        }
        if (config.rpcUser != null) {
            appendLine()
            appendLine("rpcuser=${config.rpcUser}")
        }
        if (config.rpcPassword != null) {
            appendLine("rpcpassword=${config.rpcPassword}")
        }
    }

    private fun defaultConfig(isTestnet: Boolean = false): String = serialize(Config(testnet = isTestnet))
}

File: ios/TimeCoinWallet/Network/ConfigManager.swift
import Foundation

// Parses Bitcoin-style time.conf. Mirrors Android ConfigManager exactly.
// Mainnet: {baseDir}/time.conf — Testnet: {baseDir}/testnet/time.conf
enum ConfigManager {
    private static let configFilename = "time.conf"
    private static let testnetSubdir = "testnet"
    private static let mainnetPort = 24001
    private static let testnetPort = 24101

    struct Config {
        var peers: [String] = []
        var rpcUser: String? = nil
        var rpcPassword: String? = nil
        var testnet: Bool = false
    }

    static func load(dir: URL, isTestnet: Bool = false) -> Config {
        let file = configFile(dir: dir, isTestnet: isTestnet)
        guard let content = try? String(contentsOf: file, encoding: .utf8) else {
            return Config(testnet: isTestnet)
        }
        var config = parse(content)
        config.testnet = isTestnet
        return config
    }

    static func save(dir: URL, config: Config) {
        let cfgDir = configDir(dir: dir, isTestnet: config.testnet)
        try? FileManager.default.createDirectory(at: cfgDir, withIntermediateDirectories: true)
        let file = cfgDir.appendingPathComponent(configFilename)
        try? serialize(config).write(to: file, atomically: true, encoding: .utf8)
    }

    static func readRaw(dir: URL, isTestnet: Bool = false) -> String {
        let file = configFile(dir: dir, isTestnet: isTestnet)
        return (try? String(contentsOf: file, encoding: .utf8)) ?? defaultConfig(isTestnet: isTestnet)
    }

    static func writeRaw(dir: URL, content: String, isTestnet: Bool = false) {
        let cfgDir = configDir(dir: dir, isTestnet: isTestnet)
        try? FileManager.default.createDirectory(at: cfgDir, withIntermediateDirectories: true)
        try? content.write(to: cfgDir.appendingPathComponent(configFilename), atomically: true, encoding: .utf8)
    }

    static func manualEndpoints(config: Config) -> [String] {
        let port = config.testnet ? testnetPort : mainnetPort
        return config.peers.map { peer in
            if peer.hasPrefix("http://") || peer.hasPrefix("https://") { return peer }
            if peer.contains(":") { return "http://\(peer)" }
            return "http://\(peer):\(port)"
        }
    }

    static func rpcCredentials(config: Config) -> (String, String)? {
        guard let user = config.rpcUser, let pass = config.rpcPassword else { return nil }
        return (user, pass)
    }

    // MARK: - Internal

    static func parse(_ contents: String) -> Config {
        var peers: [String] = []
        var rpcUser: String? = nil
        var rpcPassword: String? = nil

        for raw in contents.components(separatedBy: .newlines) {
            var line = raw
            if let commentIdx = line.firstIndex(of: "#") { line = String(line[..<commentIdx]) }
            line = line.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty, let eqIdx = line.firstIndex(of: "=") else { continue }
            let key = String(line[..<eqIdx]).trimmingCharacters(in: .whitespaces)
            let value = String(line[line.index(after: eqIdx)...]).trimmingCharacters(in: .whitespaces)
            guard !value.isEmpty else { continue }
            switch key {
            case "addnode": peers.append(value)
            case "rpcuser": rpcUser = value
            case "rpcpassword": rpcPassword = value
            default: break
            }
        }
        return Config(peers: peers, rpcUser: rpcUser, rpcPassword: rpcPassword)
    }

    static func serialize(_ config: Config) -> String {
        var lines = ["# TIME Coin Wallet Configuration", "# Edit this file to add masternode peers manually.", ""]
        lines.append("# Masternode peers (one per line)")
        if config.peers.isEmpty {
            let port = config.testnet ? testnetPort : mainnetPort
            lines.append("# addnode=64.91.241.10:\(port)")
        } else {
            config.peers.forEach { lines.append("addnode=\($0)") }
        }
        if let user = config.rpcUser { lines.append(""); lines.append("rpcuser=\(user)") }
        if let pass = config.rpcPassword { lines.append("rpcpassword=\(pass)") }
        return lines.joined(separator: "\n")
    }

    private static func configDir(dir: URL, isTestnet: Bool) -> URL {
        isTestnet ? dir.appendingPathComponent(testnetSubdir) : dir
    }

    private static func configFile(dir: URL, isTestnet: Bool) -> URL {
        configDir(dir: dir, isTestnet: isTestnet).appendingPathComponent(configFilename)
    }

    private static func defaultConfig(isTestnet: Bool) -> String {
        serialize(Config(testnet: isTestnet))
    }
}
